applications/nsmrs/pkg/serviceregistryserver/cache.go

```diff
@@ -0,0 +1,147 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serviceregistryserver
+
+import (
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/sirupsen/logrus"
+
+	"github.com/golang/protobuf/proto"
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/registry"
+)
+
+const (
+	// NSEExpirationTimeoutDefault - default Endpoint expiration timeout, NSE will be deleted if UpdateNetworkServiceEndpoint not received
+	NSEExpirationTimeoutDefault = 5 * time.Minute
+	// NSEExpirationTimeoutSecondsEnv - environment variable contains custom NSEExpirationTimeout
+	NSEExpirationTimeoutSecondsEnv = "NSE_EXPIRATION_TIMEOUT"
+)
+
+// NSERegistryCache - cache of registered Network Service Endpoints
+type NSERegistryCache interface {
+	AddNetworkServiceEndpoint(nse *registry.NSERegistration) (*registry.NSERegistration, error)
+	UpdateNetworkServiceEndpoint(nse *registry.NSERegistration) (*registry.NSERegistration, error)
+	DeleteNetworkServiceEndpoint(endpointName string) (*registry.NSERegistration, error)
+	GetEndpointsByNs(networkServiceName string) []*registry.NSERegistration
+	StartNSMDTracking()
+}
+
+type nseRegistryCache struct {
+	networkServiceEndpoints map[string][]*registry.NSERegistration
+	endpoints               map[string]*registry.NSERegistration
+	nseExpirationTimeout    time.Duration
+}
+
+//NewNSERegistryCache creates new nerwork service endpoints cache
+func NewNSERegistryCache() NSERegistryCache {
+	expirationTimeout := NSEExpirationTimeoutDefault
+	if interval := strings.TrimSpace(os.Getenv(NSEExpirationTimeoutSecondsEnv)); interval != "" {
+		t, err := strconv.ParseInt(interval, 10, 32)
+		if err != nil {
+			logrus.Errorf("Cannot parse %s, use default value : %v", NSEExpirationTimeoutSecondsEnv, err)
+		} else {
+			expirationTimeout = time.Duration(t) * time.Second
+		}
+	}
+
+	return &nseRegistryCache{
+		networkServiceEndpoints: make(map[string][]*registry.NSERegistration),
+		endpoints:               make(map[string]*registry.NSERegistration),
+		nseExpirationTimeout:    expirationTimeout,
+	}
+}
+
+// AddNetworkServiceEndpoint - register NSE in cache
+func (rc *nseRegistryCache) AddNetworkServiceEndpoint(entry *registry.NSERegistration) (*registry.NSERegistration, error) {
+	if endpoint, ok := rc.endpoints[entry.NetworkServiceEndpoint.Name]; ok {
```

Comment:
`rc.endpoints` is modified from different goroutines, please use some primitives of synchronization